pygmt/src/grd2xyz.py

```diff
@@ -0,0 +1,104 @@
+"""
+grd2xyz - Convert grid to data table
+"""
+import warnings
+
+import pandas as pd
+import xarray as xr
+from pygmt.clib import Session
+from pygmt.exceptions import GMTInvalidInput
+from pygmt.helpers import (
+    GMTTempFile,
+    build_arg_string,
+    fmt_docstring,
+    kwargs_to_strings,
+    use_alias,
+)
+
+
+@fmt_docstring
+@use_alias(
+    R="region",
+    V="verbose",
+)
+@kwargs_to_strings(R="sequence")
+def grd2xyz(grid, output_type="pandas", outfile=None, **kwargs):
+    r"""
+    Convert grid to data table.
+
+    Read a grid and output xyz-triplets as a :class:`numpy.ndarray`,
+    :class:`pandas.DataFrame`, or ASCII file.
+
+    Full option list at :gmt-docs:`grd2xyz.html`
+
+    {aliases}
+
+    Parameters
+    ----------
+    grid : str or xarray.DataArray
+        The file name of the input grid or the grid loaded as a
+        :class:`xarray.DataArray`. This is the only required parameter.
+    output_type : str
+        Determine the format the xyz data will be returned in [Default is
+        ``pandas``]:
+
+            - ``numpy`` - :class:`numpy.ndarray`
+            - ``pandas``- :class:`pandas.DataFrame`
+            - ``file`` - ASCII file (requires ``outfile``)
+    outfile : str
+        The file name for the output ASCII file.
+    {R}
+    {V}
+
+    Returns
+    -------
+    ret : pandas.DataFrame or numpy.ndarray or None
+        Return type depends on ``outfile`` and ``output_type``:
+
+        - None if ``outfile`` is set (output will be stored in file set by
+          ``outfile``)
+        - :class:`pandas.DataFrame` or :class:`numpy.ndarray` if ``outfile`` is
+          not set (depends on ``output_type`` [Default is
+          :class:`pandas.DataFrame`])
+
+    """
+    if output_type not in ["numpy", "pandas", "file"]:
+        raise GMTInvalidInput(
+            """Must specify format as either numpy, pandas, or file."""
+        )
+    if outfile is not None and output_type != "file":
+        msg = (
+            f"Changing `output_type` of grd2xyz from '{output_type}' to 'file' "
+            "since `outfile` parameter is set. Please use `output_type='file'` "
+            "to silence this warning."
+        )
+        warnings.warn(msg, category=RuntimeWarning, stacklevel=2)
+        output_type = "file"
+    elif output_type == "file" and outfile is None:
+        raise GMTInvalidInput("""Must specify outfile for ASCII output.""")
+    # Set the default column names for the pandas dataframe header
+    dataframe_header = ["x", "y", "z"]
+    # Set the column names to match an input DataArray as the grid
+    if isinstance(grid, xr.DataArray) and output_type == "pandas":
+        dataframe_header = grid.dims + (grid.name,)
```

Comment:
```suggestion
    if "o" not in kwargs.keys(): # Only set column names for standard output
        # Set the default column names for the pandas dataframe header
        dataframe_header = ["x", "y", "z"]
        # Set the column names to match an input DataArray as the grid
        if isinstance(grid, xr.DataArray) and output_type == "pandas":
            # Reverse the dims because it is rows, columns ordered.
            dataframe_header = (grid.dims[1],) + (grid.dims[0],) + (grid.name,)
```
Maybe support for auto-determination of columns names when `outcols` is used could be included as part of https://github.com/GenericMappingTools/pygmt/issues/1318, but for now we could not name the columns if **-o** (`outcols`) is set since we do not have a parser for string arguments and the common option is not added to this function.

Based on this code in `pygmt/clib/converstion.py`, I expect that switching the order of the dims should work for the DataArray input case.
```
    # Extract region and inc from the grid
    region = []
    inc = []
    # Reverse the dims because it is rows, columns ordered. In geographic
    # grids, this would be North-South, East-West. GMT's region and inc are
    # East-West, North-South.
    for dim in grid.dims[::-1]:
        coord = grid.coords[dim].values
        coord_incs = coord[1:] - coord[0:-1]
        coord_inc = coord_incs[0]
        if not np.allclose(coord_incs, coord_inc):
            raise GMTInvalidInput(
                f"Grid appears to have irregular spacing in the '{dim}' dimension."
            )
        if coord_inc == 0:
            raise GMTInvalidInput(
                f"Grid has a zero increment in the '{dim}' dimension."
            )
        region.extend(
            [
                coord.min() - coord_inc / 2 * grid.gmt.registration,
                coord.max() + coord_inc / 2 * grid.gmt.registration,
            ]
        )
```